### PR TITLE
Include challenge password attribute if required by EST server

### PR DESCRIPTION
**Pull request by mobe1**
The changes introduced come after the issue/feature requrest https://github.com/globalsign/est/issues/30 has been opened.
They allow us to enroll a CSR that includes the TLS-unique value as recommended by the RFC 7030

- Because each http client instantiation results in a new TLS-unique, one way of including it would be to make EST requests from the same http client.
- Because the standard crypto/x509 Go package does not handle the challenge password attribute (OID) the way an EST/CA server expects it, the CSR creation had to be wrapped.

### CODE REVIEW

**Comment by toddgaunt-gs** [comment]
Thanks for opening this PR, I'll forward this to my team for review.

**Comment by toddgaunt-gs** [comment]
Need to resolve conflicts after upgrading to Go 1.22.1...

**Comment by mobe1** [comment]
Thought the comment wasn't for me until I noticed the repo did get upgraded with a different version of go...  
The conflicts should be resolved.

**Comment by toddgaunt-gs** [comment]
Ah it wasn't actually! Just a reminder for when I had time to get around to this again, but thank you for addressing the changes needed!

**Comment by DDvO** [comment]
Looks like work on this PR has stalled - I wonder when is it going to be resumed?

**Comment by DDvO** [comment]
BTW, copying to `csr.go` large chunks of code from the `x509 package` does not look right -
there should be a lean way of adding and reading the `challengePassword` attribute of a PKCS#10 CSR.

**Comment by toddgaunt-gs** [comment]
Hi DDvO, unfortunately my team hasn't been able to spare the cycles to work on this repository for a while, so while I hesitate to recommend it you may want to fork and patch a clone of this repository if you need this functionality right away.

**Comment by toddgaunt-gs** [comment]
> BTW, copying to `csr.go` large chunks of code from the `x509 package` does not look right - there should be a lean way of adding and reading the `challengePassword` attribute of a PKCS#10 CSR.

I agree with this, I'd prefer to see this functionality without including this dependency if possible.

**Comment by toddgaunt-gs** [comment]
Sorry for such a long delay for reviewing this PR, I didn't prioritize it has been far to long so I apologize for that. Thank you for your contribution, but I don't think I can hit the merge button on this PR as it is now into the master, but I am happy to support it as a separate branch to support this use-case for now.

Some critiques that are blocking this:
- I'd prefer the smaller style changes, such as using `io` over `ioutil` and changing the code to call `EqualFold` be done in a separate PR as these changes are unrelated to the challenge password.
- As @DDvO mentions above, avoiding copying csr.go would be preferred
- The samples and resources directory added I'll have to give some more thought to. While I like seeing a way to try out the changes, I think I'd prefer this just as an automated test of the feature instead. I think adding the case to est_test.go would be the way to go.

**Comment by mobe1** [comment]
@toddgaunt-gs 
Nah acutally, you're right. I shouldn't have mixed code refactoring and the feature itself as it doesn't help telling what's necessary to make the feature available...

On a personal note: I thought I had to change the method signature but after a quick review, I start to think it can be done while keeping your methods signatures intact.

That being said,
- I'll revert those unrelated changes
- then I'll bring the method signature back
- finally, i'll see if there is a better way of adding that `tls-unique` without copying the `csr.go` (spoiler alert, i'm not very confident on this one)

**Comment by DDvO** [comment]
Hi guys,

thank you for swiftly responding on this.

> > BTW, copying to csr.go large chunks of code from the x509 package does not look right -
> > there should be a lean way of adding and reading the challengePassword attribute of a PKCS#10 CSR.
> 
> I too agree, I actually borrowed that code snippet from another repository, credit goes to [micromdm](https://github.com/globalsign/est/pull/38/files#diff-e7af9b167346dc2be28de5632d020db09219b84b9324531b935331b838745c3cR51).

Looks like there is a glitch in the link you included and instead you meant
https://github.com/micromdm/scep/blob/main/cryptoutil/x509util/x509util.go

> Meanwhile, unfortunately Go has still not taken action to address it, there is an [open issue](https://github.com/golang/go/issues/15995) for it already.

I've meanwhile had a closer look.
Oh my goodness. The `crypto/x509` package screwed up the implementation of the [PKCS#10 CSR `Attribute` type]([url](https://datatracker.ietf.org/doc/html/rfc2986#section-4)),
which is hard to work around also due to its very restrictive API.
I just [commented there](https://github.com/golang/go/issues/15995#issuecomment-2792691214) accordingly.

**Comment by mobe1** [comment]
Hey guys,

It appears Go hasn't moved much since, aside from labeling it `NeedsInvestigation`.
@DDvO, I checked out the [CSR tool](https://github.com/golang/go/issues/15995#issuecomment-2800868263) project your referenced as well
- it does add the challenge password to the CSR
- however, that's the only thing it does (most of the CSR fields are still missing)

Also noticed the project hasn't moved since its creation date and the fact he did not respond to your remark aren't very good signs.

At this point, I don't know if there are better alternatives
- I took the shortcut of using existing code from [micromdm](https://github.com/micromdm/scep/tree/main) `scep` package that already "solved" the problem, which implied a huge code duplication from `x509` package though
- Wait for Go to act on it and fix the CSR attribute implementation

File: csr.go [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+// CreateCertificateRequest creates a new certificate request based on a template.
+// The resulting CSR is similar to x509 but optionally supports the
+// challengePassword attribute.
+//
+// See https://github.com/golang/go/issues/15995
+//
+// Copied from https://github.com/micromdm/scep/blob/main/cryptoutil/x509util/x509util.go#L58
+func CreateCertificateRequest(rand io.Reader, template *CertificateRequest, priv any) (csr []byte, err error) {
+	if template == nil {
+		return nil, errors.New("make sure to rovide a certificate request template")
```
Comment (toddgaunt-gs) [inline csr.go:57]: The error string here has a typo: "make sure to _rovide_..." should be "provide"

**Review by toddgaunt-gs — CHANGES_REQUESTED** [review]
Unfortunately after further review I still don't think I'd like to merge this PR as is. Thank you for addressing prior feedback, but the following changes would need to be made before I merge this PR into the master branch.:

- Review comments addressed
- The `sample` program removed or rewritten into a directory containing a config file for the EST client and either a shell script or instructions for how to operate it for the use case it is demonstrating
- The copied challenge password code found in parts of csr.go isolated into its own package with a clear README file indicating the source without any new or custom code included alongside it.

I understand that these changes I'm asking for may not be a priority for you as you have a functional fork of this repository working for you, and updates to this one are few and far between. Hopefully this review can at least benefit your own fork if you do decide you want to pursue these changes. Let me know if so and we can discuss in more detail each point to avoid misunderstanding any of the requested changes.

**Review by mobe1 — COMMENTED** [review]
Yes, i did miss a critical detail on `PoP` in the context of `mTLS`. I'll get back to it as soon as i can.
More details on the corresponding comment.

As for the `/sample` directory, I'll just drop it cause there is no straightforward way to show it added the `challengePassword`.
I used to verify it via the EST server. It was easy and quick for me to just `go run` it.

File: client.go [hunk 2]
```diff
@@ -447,6 +485,21 @@ func (c *Client) newRequest(
...
+	if challengePassword == "" {
+		return csr, nil
+	}
+
+	standardLibCsr, _ := x509.ParseCertificateRequest(csr)
+	cr := CertificateRequest{
+		CertificateRequest: *standardLibCsr,
+		ChallengePassword:  challengePassword,
+	}
+	crBs, err := CreateCertificateRequest(rand.Reader, &cr, c.PrivateKey)
```
[the remark this answers is not in the document]
Comment (toddgaunt-gs) [reply client.go:499]: I agree, I think a new field that can be set through a command-line flag would work best. I think `SigningKey` and `signing-key` would work well respectively. Importantly if the field is nil/the flag is not provided, we should not re-sign the CSR which would result in not using the tls-unique value for the challenge-password, but I think that is preferred behavior.

**Comment by mobe1** [comment]
Hi @toddgaunt-gs, how you doing :)

Sorry for the delay, I had some time off and realized I haven't updated my PR since.

I closed the obvious comments : 
- I directly import the package that includes the tls-unique/challenge password (instead of copy/pasting with a comment).
- The sample directoy was removed as it doesn't provide any value to the repo itself.

Finally, I fixed the defect I was about to introduce, I left the comment open for reminding the context and the agreed solution.
I also updated the EST client CLI along the readme.

Let me know if something else needs to be done.

File: README.md [hunk 3]
```diff
@@ -9,18 +9,18 @@ enrollment protocol as defined by [RFC7030](https://tools.ietf.org/html/rfc7030)
...
 support for:
 
- * The optional `/csrattrs` and `/serverkeygen` operations, with support for
+* The optional `/csrattrs` and `/serverkeygen` operations, with support for
    server-generated private keys returned with or without additional
    encryption
- * The optional additional path segment
- * Optional HTTP-based client authentication on top of certificate-based
+* The optional additional path segment
+* Optional HTTP-based client authentication on top of certificate-based
```
Comment (mobe1) [inline README.md]: Juts a reminder for me -> undo this change related to single space removal in each line.

File: est_test.go [hunk 4]
```diff
@@ -381,7 +380,9 @@ func TestReenroll(t *testing.T) {
 		tc := tc
 
 		t.Run(tc.name, func(t *testing.T) {
-			// Create test EST server and client.
+			// Create test EST client.
+			client := newTestClient()
+			client.PrivateKey = tc.key
```
Comment (mobe1) [inline est_test.go]: Reminder for me -> can be reverted without any impact on unit tests since we make the difference between `private` and `signing` keys now

File: est_test.go [hunk 5]
```diff
@@ -601,6 +601,8 @@ func TestTPMEnroll(t *testing.T) {
 			// Request an EK certificate via normal enrollment.
 			ek := mustGenerateRSAPrivateKey(t)
 			csr := mustCreateCertificateRequest(t, ek, "Test TPM Device", nil)
+
+			client.PrivateKey = ek
```
Comment (mobe1) [inline est_test.go]: Reminder for me -> can be reverted without any impact on unit tests since we make the difference between `private` and `signing` keys now

**Review by mobe1 — COMMENTED** [review]
Noticed a few changes that are not longer required considering the latest solution that has been suggested/accepted.

I'll deal with them right away.

**Comment by mobe1** [comment]
@toddgaunt-gs hi,
I think i have reverted all of the unrelated changes.

I saw there hasn't been a tag yet in one of your packages.
=> I used the commit hash instead in the meantime.

And, i'm not sure if it's relevant, but just in case, I made sure I didn't break anything in the CLI.
I used different options and checked the csr `signing-key` did not interfere with the mTLS private `key`.

image

I tested with a different CA to verify the final CSR the `estclient` CLI sent did include the challenge password.

image

Tried with your testing server .exe as well, and it works just fine.
It's just I don't have a direct way of checking whether it includes any challenge password or not.

**Comment by toddgaunt-gs** [comment]
Thanks for pinging me, sorry I'd been busy and haven't gotten back around to this. I'll take a look today.

File: client.go [hunk 6]
```diff
@@ -447,6 +492,22 @@ func (c *Client) newRequest(
...
 }
 
+// addChallengePassword returns a new CSR based on the input csr.
+// The challenge password corresponds to the tls-unique value base64 encoded (only in TLS 1.2)
+func (c *Client) addChallengePassword(csr []byte, challengePassword string) ([]byte, error) {
+	if challengePassword == "" {
+		return csr, nil
+	}
+
+	stdCsr, _ := x509.ParseCertificateRequest(csr)
```
Comment (toddgaunt-gs) [inline client.go]: This error shouldn't be ignored since the CSR can originate from something the user passes in. If this fails, it may result in a panic on line 504.

File: go.sum
Comment (toddgaunt-gs) [inline go.sum:1]: Would you mind running `go tidy` on this? I see quite a few changes when I do locally, mostly removing things we don't need anymore.

**Review by toddgaunt-gs — DISMISSED** [review]

File: client.go (hunk 6 above)
Comment (mobe1) [reply client.go]: Oh, that is so true !
I didn't notice `smallstep` was using the value instead of the pointer.

Good catch, i'm on it.

File: go.sum
Comment (mobe1) [reply go.sum:1]: Huh, I was sure i did run `go mod tidy`.
Let me try again.

**Comment by toddgaunt-gs** [comment]
LGTM outside of a few small things. Feel free to address them if you have the time, though they aren't critical. I'll request another coworker to review in the meanwhile.

**Review by mobe1 — COMMENTED** [review]
Welcome back.

File: go.sum
Comment (mobe1) [reply go.sum:1]: You're right, i did miss a `tidy`...

**Review by toddgaunt-gs — APPROVED** [review]

**Comment by mobe1** [comment]
Hi @mara-soldan welcome :)

I don't know when you'll get the time to review the PR.
But here's a little briefing of what has been done if it helps
- I took advantage of the existing `makeHTTPClient` to get the `tls-unique` session/value (aka, `challenge password` in EST specs)
- I imported the `smallstep` SCEP package which handles the `challengePassword` attribute in a CSR
  - it basically wraps the standard `x509.CertificateRequest` which does not handle it correctly
  - once the [issue](https://github.com/golang/go/issues/15995) is resolved in the framework Go, we can then remove the dependency to `smallstep` custom function
- And I updated Globalsign EST client CLI as well

And of course, if you see something else, let me know and I'll address it as soon as I can.

**Comment by toddgaunt-gs** [comment]
@mobe1 merged since I didn't want to hold this up needlessly. Thanks for the contribution, and for being patient and pleasant to work with during the review/feedback cycle!

**Comment by mobe1** [comment]
@toddgaunt-gs I can say the same.
Appreciate !
